master/buildbot/process/buildstep.py

```diff
@@ -649,11 +648,12 @@ def acquireLocks(self, res=None):
                 log.msg("step %s waiting for lock %s" % (self, lock))
                 d = lock.waitUntilMaybeAvailable(self, access)
                 d.addCallback(self.acquireLocks)
-                self._acquiringLock = (lock, access, d)
+                self._acquiringLocks.append((lock, access, d))
                 return d
```

Comment:
As we return just after, I would expect this list to always say at maximum one lock.

We could have a race if waitUnrilMaybeAvailable returns succeed synchronously, then addCallback would recall acquirelocks synchronously, so _acquiringLock will be reset and restarted to lock, access, d just after, so I would switch line 651 and 652

Reply:
> As we return just after, I would expect this list to always say at maximum one lock.

Example:
The build step has two locks: `lock1` and `lock2` and neither lock is available yet. The `acquireLocks` code will `waitUntilMaybeAvailable` on `lock1`. When `lock1` becomes available, it will call the `acquireLocks` method again, and then it will have to `waitUntilMaybeAvailable` on `lock2`. At this point, we would have two items in the `self._acquiringLocks` list.

If we were to cancel this build at this point, we must clean up both `lock1` and `lock2`. The existing master code would only clean up `lock2`, so our (cancelled) build will stay in the waiters of `lock1` forever.

>We could have a race if waitUnrilMaybeAvailable returns succeed synchronously, then addCallback would recall acquirelocks synchronously, so _acquiringLock will be reset and restarted to lock, access, d just after, so I would switch line 651 and 652

Good spot, I will switch those two lines

Reply:
I'm sorry, but this explanation still does not match. At the begining of acquireLocks, we just reset the _acquiringLock variable, and we do nothing with it
acquireLocks() shall only be called once in start step, and then will be called again by waitUntilMaybeAvailable() deferred.

waitUntilMaybeAvailable() just waits, but does not actually lock the lock. That is done later with lock.claim()

This method is designed so that we check that all locks are available in a synchronous manner, and then we claim them all in a row.